rust/core-lib/src/annotations.rs

```diff
@@ -89,34 +164,193 @@ pub trait ToAnnotation {
 
 /// All the annotations for a given view
 pub struct AnnotationStore {
-    _store: HashMap<PluginId, Vec<Annotations>>,
+    store: HashMap<PluginId, Vec<Annotations>>,
 }
 
 impl AnnotationStore {
     pub fn new() -> Self {
-        AnnotationStore { _store: HashMap::new() }
+        AnnotationStore { store: HashMap::new() }
+    }
+
+    /// Invalidates and removes all annotations in the range of the interval.
+    pub fn invalidate(&mut self, interval: Interval) {
+        for val in self.store.values_mut() {
+            val.clone().into_iter().for_each(|mut r| {
+                r.invalidate(interval);
+            });
+        }
     }
 
     /// Applies an update from a plugin to a set of annotations
-    pub fn update(
-        &mut self,
-        _source: PluginId,
-        _type_id: AnnotationType,
-        _iv: Interval,
-        _items: Spans<Value>,
-    ) {
-        // todo
+    pub fn update(&mut self, source: PluginId, interval: Interval, item: Annotations) {
+        let updated_items = item.clone();
+        self.store
+            .entry(source)
+            .and_modify(|e| {
+                let mut annotations = e
+                    .iter()
+                    .filter(|a| a.annotation_type != updated_items.annotation_type)
+                    .cloned()
+                    .collect::<Vec<Annotations>>();
+
+                match e.iter_mut().find(|a| a.annotation_type == updated_items.annotation_type) {
+                    Some(outdated_annotations) => {
+                        let mut updated_annotations = outdated_annotations.clone();
+                        updated_annotations
+                            .update(Interval::new(0, interval.end()), updated_items.items);
+                        annotations.push(updated_annotations);
+                    }
+                    None => {
+                        annotations.push(updated_items);
+                    }
+                }
+
+                *e = annotations;
+            })
+            .or_insert(vec![item]);
```

Comment:
I spent a long time staring at this function trying to figure out what was going on. As far as I can tell, it's equivalent to:

```rust
if !self.store.contains_key(&source) {
    self.store.insert(source, vec![item]);
    return;
}

let entry = self.store.get_mut(&source).unwrap();
if let Some(annotation) = entry.iter_mut().find(|a| a.annotation_type == item.annotation_type) {
    annotation.update(Interval::new(0, interval.end()), item.items);
} else {
    entry.push(item);
}
```

Does this seem true to you? I'm still confused about why we zero out interval.start.

Reply:
Oh wow this is so much cleaner! Yeah I also cannot really wrap my head around the zero offset for the interval :/

Reply:
Okay, so I think this needs more thought: we're being given one interval but we're replacing it with another one. Why?

When we use spans elsewhere (such as for syntax highlighting) we ensure that the length of the total spans object is always equal to the length of the rope. In this case, the interval that we're sent by a plugin _probably_ refers to some specific region of the document, and zeroing that out would be incorrect?

This depends on the semantics of the API, so you'll have a better idea than I do at this point, but that's why this feels like it might be wrong.

Reply:
Right, good point. When we update spans eg. for syntax highlighting then the interval of the spans are all relative to the interval to be updated.
However, when updating annotations `AnnotationSlice`s are sent which specify the absolute position of the annotations in the document.

Now I'm wondering if it would be more consistent to use `Spans` instead of `AnnotationSlice` in `UpdateAnnotations`? On the other hand, there is also the `ToAnnotation` trait which could be implemented for custom types and that one returns `AnnotationSlice`. 

If we decide to keep using `AnnotationSlice` then the current problem should be easily resolved by not doing the `transform` step [that I marked in the code](https://github.com/xi-editor/xi-editor/pull/1128/files#r266140348).